File: server/backend-app/src/main/java/com/sawwere/titlecounter/backend/app/storage/entity/Game.java
package com.sawwere.titlecounter.backend.app.storage.entity;

import jakarta.persistence.AttributeOverride;
import jakarta.persistence.AttributeOverrides;
import jakarta.persistence.Column;
import jakarta.persistence.Embedded;
import jakarta.persistence.Entity;
import jakarta.persistence.FetchType;
import jakarta.persistence.GeneratedValue;
import jakarta.persistence.GenerationType;
import jakarta.persistence.Id;
import jakarta.persistence.JoinColumn;
import jakarta.persistence.JoinTable;
import jakarta.persistence.ManyToMany;
import jakarta.persistence.OneToMany;
import jakarta.persistence.Table;
import java.time.LocalDate;
import java.time.LocalDateTime;
import java.util.ArrayList;
import java.util.List;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;
import org.hibernate.annotations.ColumnDefault;
import org.hibernate.annotations.CreationTimestamp;
import org.hibernate.annotations.UpdateTimestamp;


@Entity
@Getter
@Setter
@Builder
@NoArgsConstructor
@AllArgsConstructor
@Table(name = "games")
public class Game {
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private long id;

    @Column(name = "title", nullable = false)
    private String title;

    @Column(name = "time")
    private long time;

    @Column(name = "date_release")
    private LocalDate dateRelease;

    @Column(name = "global_score")
    private Float globalScore;

    @Column(name = "game_type")
    private String gameType;

    @Column(name  = "developer")
    private String developer;

    @Column(name = "description", length = 2048)
    private String description;

    @Embedded
    @AttributeOverrides({
            @AttributeOverride(name = "hltbId", column = @Column(name = "hltb_id", unique = true)),
            @AttributeOverride(name = "steamId", column = @Column(name = "steam_id", unique = true))
    })
    private GameExternalId externalId;

    @ManyToMany(fetch = FetchType.EAGER)
    @JoinTable(
            name = "games_game_platforms",
            joinColumns = @JoinColumn(name = "game_id"),
            inverseJoinColumns = @JoinColumn(name = "game_platform_id")
    )
    @Builder.Default
    private List<GamePlatform> platforms = new ArrayList<>();

    @ManyToMany(fetch = FetchType.EAGER)
    @JoinTable(
            name = "games_game_genres",
            joinColumns = @JoinColumn(name = "game_id"),
            inverseJoinColumns = @JoinColumn(name = "game_genre_id")
    )
    @Builder.Default
    private List<GameGenre> genres = new ArrayList<>();

    @ManyToMany(fetch = FetchType.EAGER)
    @JoinTable(
            name = "games_game_developers",
            joinColumns = @JoinColumn(name = "game_id"),
            inverseJoinColumns = @JoinColumn(name = "game_developer_id")
    )
    @Builder.Default
    private List<GameDeveloper> developers = new ArrayList<>();

    @OneToMany(orphanRemoval = true, mappedBy = "game")
    private List<GameEntry> gameEntries;

    @Column(name = "created_at", nullable = false)
    @ColumnDefault("'2024-08-04 10:23:54'::timestamp without time zone")
    @CreationTimestamp
    private LocalDateTime createdAt;

    @Column(name = "updated_at", nullable = false)
    @ColumnDefault("'2024-08-04 10:23:54'::timestamp without time zone")
    @UpdateTimestamp
    private LocalDateTime  updatedAt;
}

File: external-api-service/models/Game.py
class Game:
    def __init__(self,
                 platforms,
                 genres,
                 hltb_id=None,
                 steam_id=None,
                 title="ERROR_MISSING_TITLE",
                 game_type="game",
                 developer=None,
                 description=None,
                 image_url=None,
                 time=None,
                 date_release=None,
                 score=0,
                 similarity=0.0
                 ):
        self.external_id = {"hltb_id": hltb_id, "steam_id": steam_id}
        self.title = title
        self.game_type = game_type
        self.developer = developer
        self.description = description
        self.image_url = image_url       

        mt = str(time)
        if (time is str):
            if ("\u00bd" in mt):
                self.time = int(mt[:-1])+0.5
            else:
                self.time = int(mt)
        else:
            self.time = time

        self.date_release = date_release
        self.score = score
        self.similarity = similarity
        self.platforms = platforms
        self.genres = genres

    def to_dict(self):
        return {
                "title": self.title,
                "game_type": self.game_type,
                "developer": self.developer,
                "description": self.description,
                "external_id": self.external_id,
                "platforms": self.platforms,
                "genres": self.genres,
                "time": self.time,
                "date_release": self.date_release,
                "global_score": self.score,
                "image_url": self.image_url
                }
